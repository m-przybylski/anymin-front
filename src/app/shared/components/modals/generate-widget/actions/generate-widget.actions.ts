import { Action } from '@ngrx/store';

export enum GenerateWidgetActionTypes {
  OpenGenerateWidgetModal = '[Generate Widget] Open generate widget',
}

export class StartOpenGenerateWidgetAction implements Action {
  public readonly type = GenerateWidgetActionTypes.OpenGenerateWidgetModal;
  constructor(public payload: { expertId?: string; serviceId: string; shareLink: string }) {}
}

export type GenerateWidgetActionsUnion = StartOpenGenerateWidgetAction;
