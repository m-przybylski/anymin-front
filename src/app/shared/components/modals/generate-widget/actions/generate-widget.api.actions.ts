import { Action } from '@ngrx/store';

export enum GenerateWidgetApiActionTypes {
  FetchWidgetIdSuccess = '[Generate Widget] Fetch Widget Id Success',
  FetchWidgetIdFailure = '[Generate Widget] Fetch Widget Id Failure',
}
// tslint:disable:max-classes-per-file

export class FetchWidgetIdSuccessAction implements Action {
  public readonly type = GenerateWidgetApiActionTypes.FetchWidgetIdSuccess;

  constructor(public payload: { expertId?: string; serviceId: string; widgetId: string }) {}
}

export class FetchWidgetIdErrorAction implements Action {
  public readonly type = GenerateWidgetApiActionTypes.FetchWidgetIdFailure;

  // tslint:disable-next-line:no-any
  constructor(public error: any) {}
}

export type GenerateWidgetActionsUnion = FetchWidgetIdSuccessAction | FetchWidgetIdErrorAction;
