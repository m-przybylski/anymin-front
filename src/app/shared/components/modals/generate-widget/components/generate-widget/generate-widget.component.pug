plat-modal-component(
  modalTrTitleHeader='GENERATE_WIDGET_MODAL.HEADER',
  [isBackwardVisible]="stepper.selectedIndex !== 0",
  (backwardClick)='stepper.previous()'
  )
  plat-stepper(contentHeightAnimation, #stepper='platStepper')
    plat-step
      .generate-widget-step-content
        .generate-widget-step-content-panel
          .generate-widget-step-content-section
            p(innerHtml= "{{'GENERATE_WIDGET_MODAL.CONTENT.TEXT1' | translate}}")
            .generate-widget-copy([plat-copy-to-clipboard]="widgetLink") {{widgetLink}}
            .generate-widget-social-share
              .generate-widget-social-link(*ngFor='let icon of socialMediaLinks')
                button(plat-icon-button)
                  span(plat-icon, [iconClass]='icon.iconName')
          .generate-widget-step-content-section
            p(innerHtml="{{'GENERATE_WIDGET_MODAL.CONTENT.TEXT2' | translate}}")
            .generate-widget-copy([plat-copy-to-clipboard]="widgetId") {{ widgetId }}
          .generate-widget-step-content-section
            p(innerHtml="{{'GENERATE_WIDGET_MODAL.CONTENT.TEXT3' | translate}}")
            button(plat-stroked-button, (click)='stepper.next()') {{ 'GENERATE_WIDGET_MODAL.CONTENT.INSTALL' | translate}}
        .generate-widget-step-content-panel-separator
        .generate-widget-step-content-panel
          .preview-wrapper
            iframe.widget-preview(
              [src]="iframeSrc",
              (load)="iframeLoaded()"
              )
            .widget-preview__loader(*ngIf="isIframeLoading")
              span(plat-icon, iconClass='loading')
    plat-step
      .generate-widget-step-content
        .generate-widget-step-content-panel
          .generate-widget-step-content-section
            p {{"GENERATE_WIDGET_MODAL.CONTENT.ADD_STEP1" | translate}}
            plat-generate-widget-button-type([formControl]='buttonType')
          .generate-widget-step-content-section(
            *ngIf="selectedButtonType !== undefinedSelectedButtonType",
            [@fadeInOut],
            (@fadeInOut.done)="animationDone($event)"
            )
            p {{"GENERATE_WIDGET_MODAL.CONTENT.ADD_STEP2" | translate}}
            .generate-widget-copy([plat-copy-to-clipboard]="headScript") {{headScript}}
          .generate-widget-step-content-section(*ngIf="displayButtonCode", [@fadeInOut])
            p {{"GENERATE_WIDGET_MODAL.CONTENT.ADD_STEP3" | translate}}
            .generate-widget-copy([plat-copy-to-clipboard]="buttonCode") {{buttonCode}}


