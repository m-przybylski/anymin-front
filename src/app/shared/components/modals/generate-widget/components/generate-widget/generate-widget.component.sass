@import "~@anymind-ng/core/src/template/stylesheets/core/core"
@import "../../../../../../template/core/core"

.generate-widget
  &-step-content
    display: flex
    flex-direction: column

    +smm
      flex-direction: row
      min-height: 687px

    &-panel
      flex: 1 0 0

      &-separator
        height: 40px

        +smm
          flex-basis: 40px

    &-section
      margin: 0 0 48px 0

    &-section:nth-last-child(1)
      margin: 0

  &-social-share
    display: flex

  &-social-link
    margin: 0 0 0 20px

  &-social-link:nth-child(1)
    margin: 0

  &-copy
    +subtitle-small
    cursor: pointer
    color: $c-text-secondary
    background: $c-base-background
    border-radius: 4px
    border: 1px dashed $c-lines
    padding: 16px
    margin: 0 0 24px 0

    &:only-of-type
      margin: 0


.preview-wrapper
  position: relative
  border-radius: 12px
  background: $c-base-background
  padding: 4px
  width: 100%
  height: 100%

.widget-preview
  width: 100%
  height: 100%

  &__loader
    position: absolute
    top: 0
    left: 0
    width: 100%
    height: 100%
    display: flex
    justify-content: center
    align-items: center

    span
      display: block
      width: 40px
      height: 40px
      font-size: 40px
      color: $c-primary
      animation: rotate360Deg 0.6s linear infinite

p
  +button
  color: $c-text-primary
  margin: 0 0 24px 0
