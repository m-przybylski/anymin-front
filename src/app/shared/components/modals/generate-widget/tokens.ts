import { InjectionToken } from '@angular/core';

export const GENERATE_WIDGET_DATA = new InjectionToken<IGenerateWidgetData>('Generate Widget Data Token');

export interface IGenerateWidgetData {
  expertId: string;
  serviceId: string;
  widgetId: string;
}
