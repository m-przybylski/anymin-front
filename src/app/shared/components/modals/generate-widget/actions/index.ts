import * as GenerateWidgetActions from './generate-widget.actions';
import * as GenerateWidgetApiActions from './generate-widget.api.actions';
export { GenerateWidgetActions, GenerateWidgetApiActions };
