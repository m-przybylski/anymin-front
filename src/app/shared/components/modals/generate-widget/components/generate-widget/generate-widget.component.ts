import { Component, Inject, AfterViewInit, ViewChild, OnDestroy, OnInit } from '@angular/core';
import { Logger } from '@platform/core/logger';
import { LoggerFactory } from '@anymind-ng/core';
import { GENERATE_WIDGET_DATA, IGenerateWidgetData } from '../../tokens';
import { ModalComponent } from '@platform/shared/components/modals/modal/modal.component';
import { GenerateWidgetDataService } from '../../services/generate-widget.data.service';
import { of, BehaviorSubject, Subject } from 'rxjs';
import { AccountPresenceStatus } from '@anymind-ng/api';
import { finalize, filter, first, takeUntil, switchMap, map } from 'rxjs/operators';
import { FormControl } from '@angular/forms';
import { WidgetButtonType } from '../generate-widget-button-type/generate-widget-button-type.component';
import { trigger, transition, style, animate, state, AnimationEvent } from '@angular/animations';

interface IShareLink {
  url: string;
  iconName: string;
}
@Component({
  selector: 'plat-generate-widget',
  templateUrl: 'generate-widget.component.html',
  styleUrls: ['generate-widget.component.sass'],
  animations: [
    trigger('fadeInOut', [
      state('void', style({ opacity: 0 })),
      transition(':enter', [animate('250ms', style({ opacity: '1' }))]),
      transition(':leave', [animate('250ms')]),
    ]),
  ],
})
export class GenerateWidgetComponent extends Logger implements OnInit, AfterViewInit, OnDestroy {
  public serviceId: string;
  public expertId: string;
  public widgetLink: string;
  public widgetId: string;
  public headScript: string;
  public avatarToken: string;
  public expertName: string;
  public consultationHeader: string;
  public cosultationDescription: string;
  public consultationPrice: string;
  public buttonCode = '';

  public readonly available = of(AccountPresenceStatus.StatusEnum.Available);
  public readonly undefinedSelectedButtonType = -1;
  public readonly socialMediaLinks: ReadonlyArray<IShareLink> = [
    { url: '', iconName: 'linkedin' },
    { url: '', iconName: 'instagram' },
    { url: '', iconName: 'facebook' },
  ];

  public buttonType: FormControl;
  public selectedButtonType: WidgetButtonType = this.undefinedSelectedButtonType;

  @ViewChild(ModalComponent)
  private modal: ModalComponent;

  /**
   * used to show if component loaded data from the server
   */
  private componentLoaded = new BehaviorSubject<boolean>(false);
  /**
   * used to trigger when component is destroyed
   */
  private destroyed$ = new Subject<void>();
  /**
   * used for animations. * => void is completed emits value
   */
  private fadeOutComplete$ = new Subject<void>();

  constructor(
    private generateWidgetDataService: GenerateWidgetDataService,
    @Inject(GENERATE_WIDGET_DATA) data: IGenerateWidgetData,
    loggerFactory: LoggerFactory,
  ) {
    super(loggerFactory.createLoggerService('GenerateWidgetComponent'));
    this.serviceId = data.serviceId;
    this.expertId = data.expertId;
    this.widgetId = data.widgetId;
  }
  public ngOnInit(): void {
    this.widgetLink = this.generateWidgetDataService.getWidgetLink(this.widgetId);
    this.headScript = this.generateWidgetDataService.getWidgetSdkLink(this.widgetId);
    this.generateWidgetDataService
      .resolve(this.widgetId)
      .pipe(
        finalize(() => {
          this.componentLoaded.next(true);
        }),
      )
      .subscribe(generateWidgetResolveData => {
        if (generateWidgetResolveData) {
          this.avatarToken = generateWidgetResolveData.expertAvatar || '';
          this.expertName = generateWidgetResolveData.expertName || '';
          this.consultationHeader = generateWidgetResolveData.serviceName;
          this.cosultationDescription = generateWidgetResolveData.serviceDesc;
          this.consultationPrice = generateWidgetResolveData.servicePrice;
        }
      });
    this.buttonType = new FormControl('');
    /**
     * once value is changes in the form
     * wait for animation to finish
     * and than start new animation.
     * avoid jumping between states and softer
     * content change
     */
    this.buttonType.valueChanges
      .pipe(
        takeUntil(this.destroyed$),
        switchMap(buttonType => {
          if (this.selectedButtonType === this.undefinedSelectedButtonType) {
            return of(buttonType);
          }
          this.selectedButtonType = this.undefinedSelectedButtonType;

          return this.fadeOutComplete$.pipe(map(() => buttonType));
        }),
      )
      .subscribe(buttonType => {
        this.selectedButtonType = buttonType;
        this.headScript = this.generateWidgetDataService.getWidgetSdkLink(
          buttonType === WidgetButtonType.FLOATING ? this.widgetId : undefined,
        );
        this.buttonCode = this.generateWidgetDataService.getButtonCode(this.widgetId, buttonType);
      });
  }
  public ngAfterViewInit(): void {
    /**
     * before stopping loader need to wait for data to be loaded
     */
    this.componentLoaded
      .pipe(
        filter(loaded => loaded),
        first(),
      )
      .subscribe(() => {
        this.modal.stopLoadingAnimation();
      });
  }
  public ngOnDestroy(): void {
    this.destroyed$.next();
    this.destroyed$.complete();
  }
  public animationDone(animationEvent: AnimationEvent): void {
    if (animationEvent.toState === 'void') {
      this.fadeOutComplete$.next();
    }
  }
  public get displayButtonCode(): boolean {
    return this.selectedButtonType === WidgetButtonType.BANNER || this.selectedButtonType === WidgetButtonType.STATIC;
  }
}
