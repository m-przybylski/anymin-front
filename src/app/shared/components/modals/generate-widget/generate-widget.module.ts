import { NgModule } from '@angular/core';
import { GenerateWidgetComponent } from './components/generate-widget/generate-widget.component';
import { GenerateWidgetService } from './services/generate-widget.service';
import { GenerateWidgetDataService } from './services/generate-widget.data.service';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import { reducer } from './reducers/generate-widget.reducer';
import { GenerateWidgetEffects } from './effects/generate-widget.effects';
import { ModalComponentsModule } from '@platform/shared/components/modals/modal/modal.components.module';
import { StepperModule } from '@platform/shared/components/stepper/stepper.module';
import { ButtonModule, IconModule } from '@platform/shared/components/atomic-components';
import { CommonModule } from '@angular/common';
import { TranslateModule } from '@ngx-translate/core';
import { UserAvatarModule } from '@platform/shared/components/user-avatar/user-avatar.module';
import { ExpertAvailabilityModule } from '@platform/features/dashboard/components/expert-availability/expert-availability.module';
import { GenerateWidgetButtonTypeComponent } from './components/generate-widget-button-type/generate-widget-button-type.component';
import { ReactiveFormsModule } from '@angular/forms';
import { CopyCodeToClipbord } from './components/copy-generated-code/copy-generated-code.component';

@NgModule({
  declarations: [GenerateWidgetComponent, GenerateWidgetButtonTypeComponent, CopyCodeToClipbord],
  imports: [
    CommonModule,
    ModalComponentsModule,
    ButtonModule,
    IconModule,
    StepperModule,
    UserAvatarModule,
    ReactiveFormsModule,
    StoreModule.forFeature('generateWidget', reducer),
    EffectsModule.forFeature([GenerateWidgetEffects]),
    TranslateModule.forChild(),
    ExpertAvailabilityModule,
  ],
  providers: [GenerateWidgetService, GenerateWidgetDataService],
  entryComponents: [GenerateWidgetComponent],
})
export class GenerateWidgetModule {}
