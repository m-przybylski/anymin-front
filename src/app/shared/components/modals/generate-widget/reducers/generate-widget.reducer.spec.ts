import { IGenerateWidgetState, reducer } from './generate-widget.reducer';
import { GenerateWidgetActions, GenerateWidgetApiActions } from '../actions';

describe('generate-widget.reducer', () => {
  const initialState: IGenerateWidgetState = {
    fetchWidgetIdPending: false,
  };
  describe('default state', () => {
    it('should apply default state', () => {
      const result = reducer(undefined, {} as any);
      expect(result).toEqual(initialState);
    });
  });
  describe('StartOpenGenerateWidgetModal', () => {
    const action = new GenerateWidgetActions.StartOpenGenerateWidgetAction({
      expertId: 'fake expert id',
      serviceId: 'fake service id',
      shareLink: 'fake url',
    });
    it('should populate state with expert and service', () => {
      const result = reducer(initialState, action);
      expect(result).toMatchSnapshot();
    });
  });
  describe('FetchWidgetIdSuccess', () => {
    const action = new GenerateWidgetApiActions.FetchWidgetIdSuccessAction({
      expertId: 'fake expert Id',
      serviceId: 'fake service Id',
      widgetId: 'fake widget Id',
    });
    it('should update store with given value', () => {
      const result = reducer({ ...initialState, fetchWidgetIdPending: true }, action);
      expect(result).toMatchSnapshot();
    });
  });
  describe('FetchWidgetIdError', () => {
    const action = new GenerateWidgetApiActions.FetchWidgetIdErrorAction('fake error');
    it('should update store with given value', () => {
      const result = reducer({ ...initialState, fetchWidgetIdPending: true }, action);
      expect(result).toMatchSnapshot();
    });
  });
});
