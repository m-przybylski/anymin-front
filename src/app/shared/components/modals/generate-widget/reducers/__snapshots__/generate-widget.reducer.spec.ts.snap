// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate-widget.reducer FetchWidgetIdError should update store with given value 1`] = `
Object {
  "fetchWidgetIdPending": false,
}
`;

exports[`generate-widget.reducer FetchWidgetIdSuccess should update store with given value 1`] = `
Object {
  "expertId": "fake expert Id",
  "fetchWidgetIdPending": false,
  "serviceId": "fake service Id",
  "widgetId": "fake widget Id",
}
`;

exports[`generate-widget.reducer StartOpenGenerateWidgetModal should populate state with expert and service 1`] = `
Object {
  "fetchWidgetIdPending": true,
  "shareUrl": "fake url",
}
`;
